### PR TITLE
Issue 602

### PR DESCRIPTION
**Pull request by kieranomahony98**
Flow:

Check if docker network is running. 
If network is not running exit
If running, and containers are present + certs are present skip network up
If running, containers are not present then we should networkUp
If containers exist but certs don’t, bring network down and then networkUp

When containers are running but for some reason peerOrganisation dir was not present, the following error: 
`Failed to generate channel configuration transaction...` may be awkward to debug. So this is why there is a check to see if peerOrg does not exist but the containers do. 


Test Cases: 

1) Docker is not running 
```
./network.sh up createChannel -ca
Using docker and docker-compose
Creating channel 'mychannel'.
If network is not up, starting nodes with CLI timeout of '5' tries and CLI delay of '3' seconds and using database 'leveldb 
docker network is required to be running to create a channel
```

2) Docker is running but containers are not, peerOrganizations directory exists
```
./network.sh up createChannel -ca
Using docker and docker-compose
Creating channel 'mychannel'.
If network is not up, starting nodes with CLI timeout of '5' tries and CLI delay of '3' seconds and using database 'leveldb 
Bringing up network
LOCAL_VERSION=2.4.1
DOCKER_IMAGE_VERSION=2.4.1
CA_LOCAL_VERSION=1.5.2
CA_DOCKER_IMAGE_VERSION=1.5.2
WARN[0000] Found orphan containers ([ca_org2 ca_org1 ca_orderer]) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up. 
[+] Running 4/4
 ⠿ Container peer0.org2.example.com  Started                                                                                                                                                                                                                               0.6s
 ⠿ Container peer0.org1.example.com  Started                                                                                                                                                                                                                               0.6s
 ⠿ Container orderer.example.com     Started                                                                                                                                                                                                                               0.6s
 ⠿ Container cli                     Started  
```
3) Containers running but  peerOrganizations directory does not exist
```
./network.sh up createChannel -ca
Using docker and docker-compose
Creating channel 'mychannel'.
If network is not up, starting nodes with CLI timeout of '5' tries and CLI delay of '3' seconds and using database 'leveldb with crypto from 'Certificate Authorities'
Decomposing compose-test-net.yaml
[+] Running 10/7
 ⠿ Container orderer.example.com          Removed                                                                                                                                                                                                                          0.2s
 ⠿ Container cli                          Removed                                                                                                                                                                                                                         10.2s
 ⠿ Container peer0.org1.example.com       Removed                                                                                                                                                                                                                          0.3s
 ⠿ Container peer0.org2.example.com       Removed                                                                                                                                                                                                                          0.3s
 ⠿ Container ca_orderer                   Removed                                                                                                                                                                                                                          0.3s
 ⠿ Container ca_org2                      Removed                                                                                                                                                                                                                          0.3s
 ⠿ Container ca_org1                      Removed                                                                                                                                                                                                                          0.3s
 ⠿ Volume compose_peer0.org2.example.com  Removed                                                                                                                                                                                                                          0.0s
 ⠿ Volume compose_peer0.org1.example.com  Removed                                                                                                                                                                                                                          0.0s
 ⠿ Volume compose_orderer.example.com     Removed                                                                                                                                                                                                                          0.0s
Decomposing compose-couch.yaml
service "peer0.org1.example.com" has neither an image nor a build context specified: invalid compose project
Decomposing compose-ca.yaml
[+] Running 1/0
 ⠿ compose  Warning: No resource found to remove                                                                                                                                                                                                                           0.0s
Error: No such volume: docker_orderer.example.com
Error: No such volume: docker_peer0.org1.example.com
Error: No such volume: docker_peer0.org2.example.com
Removing remaining containers
Removing generated chaincode docker images
"docker kill" requires at least 1 argument.
See 'docker kill --help'.

Usage:  docker kill [OPTIONS] CONTAINER [CONTAINER...]

Kill one or more running containers
Bringing up network
LOCAL_VERSION=2.4.1
DOCKER_IMAGE_VERSION=2.4.1
CA_LOCAL_VERSION=1.5.2
...
```

Signed-off-by: Kieran O Mahony [Kieran.O.Mahony1@ibm.com](mailto:Kieran.O.Mahony1@ibm.com)

### CODE REVIEW

File: test-network/network.sh [hunk 1]
```diff
@@ -267,9 +267,22 @@ function networkUp() {
...
+    fatalln "$CONTAINER_CLI network is required to be running to create a channel"
+  fi
+
+  # check if all containers are present 
+  CONTAINERS=($($CONTAINER_CLI ps | grep hyperledger/ | awk '{print $2}'))
+  len=$(echo ${#CONTAINERS[@]})
+
+  [ $len -eq 7 ] && [ ! -d "organizations/peerOrganizations" ] && networkDown
+
+  [ $len -lt 7 ] || [ ! -d "organizations/peerOrganizations" ] && bringUpNetwork="true" || echo "Network Running Already"
```
Comment (nikhil550) [inline test-network/network.sh:283]: when would `bringUpNetwork="true"` be set. I only see it being set one time as false.

**Review by nikhil550 — COMMENTED** [review]
When I do the `network.sh createChannel` command with the network up, it still tries to bring up the network.

```
docker ps 
CONTAINER ID   IMAGE                               COMMAND             CREATED              STATUS              PORTS                                                                    NAMES
597ec2e4b24a   hyperledger/fabric-tools:latest     "/bin/bash"         About a minute ago   Up About a minute                                                                            cli
bdacd602135a   hyperledger/fabric-peer:latest      "peer node start"   About a minute ago   Up About a minute   0.0.0.0:7051->7051/tcp, 0.0.0.0:9444->9444/tcp                           peer0.org1.example.com
6b17408a87b9   hyperledger/fabric-peer:latest      "peer node start"   About a minute ago   Up About a minute   0.0.0.0:9051->9051/tcp, 7051/tcp, 0.0.0.0:9445->9445/tcp                 peer0.org2.example.com
6f4a5cdcd2cf   hyperledger/fabric-orderer:latest   "orderer"           About a minute ago   Up About a minute   0.0.0.0:7050->7050/tcp, 0.0.0.0:7053->7053/tcp, 0.0.0.0:9443->9443/tcp   orderer.example.com

./network.sh up createChannel -c channel1
Using docker and docker-compose
Creating channel 'channel1'.
If network is not up, starting nodes with CLI timeout of '5' tries and CLI delay of '3' seconds and using database 'leveldb 
Bringing up network
LOCAL_VERSION=2.4.2
DOCKER_IMAGE_VERSION=2.4.2
[+] Running 4/4
 ⠿ Container peer0.org1.example.com  Running                               0.0s
 ⠿ Container peer0.org2.example.com  Running                               0.0s
 ⠿ Container orderer.example.com     Ru...                                 0.0s
 ⠿ Container cli                     Running                               0.0s
CONTAINER ID   IMAGE                               COMMAND             CREATED         STATUS         PORTS                                                                    NAMES
597ec2e4b24a   hyperledger/fabric-tools:latest     "/bin/bash"         2 minutes ago   Up 2 minutes                                                                            cli
bdacd602135a   hyperledger/fabric-peer:latest      "peer node start"   2 minutes ago   Up 2 minutes   0.0.0.0:7051->7051/tcp, 0.0.0.0:9444->9444/tcp                           peer0.org1.example.com
6b17408a87b9   hyperledger/fabric-peer:latest      "peer node start"   2 minutes ago   Up 2 minutes   0.0.0.0:9051->9051/tcp, 7051/tcp, 0.0.0.0:9445->9445/tcp                 peer0.org2.example.com
6f4a5cdcd2cf   hyperledger/fabric-orderer:latest   "orderer"           2 minutes ago   Up 2 minutes   0.0.0.0:7050->7050/tcp, 0.0.0.0:7053->7053/tcp, 0.0.0.0:9443->9443/tcp   orderer.example.com
```

File: test-network/network.sh [hunk 2]
```diff
@@ -267,9 +267,22 @@ function networkUp() {
...
-  if [ ! -d "organizations/peerOrganizations" ]; then
+  if ! $CONTAINER_CLI info > /dev/null 2>&1 ; then
+    fatalln "$CONTAINER_CLI network is required to be running to create a channel"
+  fi
+
+  # check if all containers are present 
+  CONTAINERS=($($CONTAINER_CLI ps | grep hyperledger/ | awk '{print $2}'))
+  len=$(echo ${#CONTAINERS[@]})
+
+  [ $len -eq 7 ] && [ ! -d "organizations/peerOrganizations" ] && networkDown
```
[the remark this answers is not in the document]
Comment (kieranomahony98) [reply test-network/network.sh:281]: When I run network up I get:
```
docker ps 
CONTAINER ID   IMAGE                               COMMAND                  CREATED         STATUS         PORTS                                                                    NAMES
4baf233a9ae6   hyperledger/fabric-tools:latest     "/bin/bash"              2 minutes ago   Up 2 minutes                                                                            cli
fe20ef2fca7b   hyperledger/fabric-peer:latest      "peer node start"        2 minutes ago   Up 2 minutes   0.0.0.0:9051->9051/tcp, 7051/tcp, 0.0.0.0:9445->9445/tcp                 peer0.org2.example.com
f5fa4bebd30e   hyperledger/fabric-peer:latest      "peer node start"        2 minutes ago   Up 2 minutes   0.0.0.0:7051->7051/tcp, 0.0.0.0:9444->9444/tcp                           peer0.org1.example.com
c1ec134133f2   hyperledger/fabric-orderer:latest   "orderer"                2 minutes ago   Up 2 minutes   0.0.0.0:7050->7050/tcp, 0.0.0.0:7053->7053/tcp, 0.0.0.0:9443->9443/tcp   orderer.example.com
b1470f0f97fb   hyperledger/fabric-ca:latest        "sh -c 'fabric-ca-se…"   3 minutes ago   Up 3 minutes   0.0.0.0:8054->8054/tcp, 7054/tcp, 0.0.0.0:18054->18054/tcp               ca_org2
b28b06714ae2   hyperledger/fabric-ca:latest        "sh -c 'fabric-ca-se…"   3 minutes ago   Up 2 minutes   0.0.0.0:9054->9054/tcp, 7054/tcp, 0.0.0.0:19054->19054/tcp               ca_orderer
a19c1b80aa75   hyperledger/fabric-ca:latest        "sh -c 'fabric-ca-se…"   3 minutes ago   Up 2 minutes   0.0.0.0:7054->7054/tcp, 0.0.0.0:17054->17054/tcp                         ca_org1
```
Can the number of containers vary?

File: test-network/network.sh (hunk 1 above)
Comment (kieranomahony98) [reply test-network/network.sh:283]: `bringUpNetwork` should be set to true if either conditional on 283 does not `exit 0`

```
  [ $len -lt 7 ] || [ ! -d "organizations/peerOrganizations" ] && bringUpNetwork="true" || echo "Network Running Already"
```

File: test-network/network.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (kieranomahony98) [reply test-network/network.sh]: Updated now

File: test-network/network.sh [hunk 3]
```diff
@@ -267,9 +267,22 @@ function networkUp() {
...
+  if ! $CONTAINER_CLI info > /dev/null 2>&1 ; then
+    fatalln "$CONTAINER_CLI network is required to be running to create a channel"
+  fi
+
+  # check if all containers are present 
+  CONTAINERS=($($CONTAINER_CLI ps | grep hyperledger/ | awk '{print $2}'))
+  len=$(echo ${#CONTAINERS[@]})
+  echo "len: $len"
+  [[ $len -ge 4 ]] && [[ ! -d "organizations/peerOrganizations" ]]  &&  networkDown 
+
```
Comment (nikhil550) [inline test-network/network.sh:282]: Its a nit, but you may want to display a message that you are bringing the network or any running containers down here if you are running the down command.

**Review by nikhil550 — APPROVED** [review]

**Comment by nikhil550** [comment]
Ran with the updates and everything worked as intended.

File: test-network/network.sh (hunk 3 above)
Comment (kieranomahony98) [reply test-network/network.sh:285]: Added in echo

**Review by lehors — APPROVED** [review]
LGTM but one of your commit didn't have a sign-off so the DCO test is failing. You can either add the sign-off on that commit or simply squash all your commit into one with the sign-off.
Thanks.
The other errors are not related to this change.

**Comment by kieranomahony98** [comment]
@lehors DCO update now thank you

**Review by nikhil550 — APPROVED** [review]

**Pull request merged**